File: lib/rgfa/line/edge/link/equivalence.rb
module RGFA::Line::Edge::Link::Equivalence

  # Computes an hash for including a link in an Hash tables,
  # so that the hash of a link and its complement is the same.
  # Thereby, tags are not considered.
  # @see #eql?
  def hash
    from_end.hash + to_end.hash + overlap.hash + complement_overlap.to_s.hash
  end

  # Compares two links and determine their equivalence.
  # Thereby, tags are not considered.
  #
  # @note Inverting the strand of both links and reversing
  #   the CIGAR operations (order/type), one obtains an
  #   equivalent complement link.
  #
  # @param other [RGFA::Line::Edge::Link] a link
  # @return [Boolean] are self and other equivalent?
  # @see #==
  # @see #same?
  # @see #complement?
  def eql?(other)
    same?(other) or complement?(other)
  end

  # Compares the tags of two links.
  #
  # @note This method shall be overridden if custom tags
  #   are defined, which have a ``complementation'' operation which determines
  #   their value in the equivalent but complement link.
  #
  # @param other [RGFA::Line::Edge::Link] a link
  # @return [Boolean] are self and other equivalent?
  # @see #==
  def eql_tags?(other)
    (self.tagnames.sort == other.tagnames.sort) and
      tagnames.each {|fn| self.get(fn) == other.get(fn)}
  end

  # Compares two links and determine their equivalence.
  # Tags must have the same content.
  #
  # @note Inverting the strand of both links and reversing
  #   the CIGAR operations (order/type), one obtains an equivalent
  #   link.
  #
  # @param other [RGFA::Line::Edge::Link] a link
  # @return [Boolean] are self and other equivalent?
  # @see #eql?
  # @see #eql_tags?
  #def ==(other)
  #  eql?(other) and eql_tags?(other)
  #end

  # Compares two links and determine their equivalence.
  # Thereby, tags are not considered.
  #
  # @param other [RGFA::Line::Edge::Link] a link
  # @return [Boolean] are self and other equivalent?
  # @see #eql?
  # @see #complement?
  # @see #==
  def same?(other)
    (from_end == other.from_end and
      to_end == other.to_end and
      overlap == other.overlap)
  end

  # Compares the link to the complement of another link
  # and determine their equivalence.
  # Thereby, tags are not considered.
  #
  # @param other [RGFA::Line::Edge::Link] the other link
  # @return [Boolean] are self and the complement of other equivalent?
  # @see #eql?
  # @see #same?
  # @see #==
  def complement?(other)
    (from_end == other.to_end and
      to_end == other.from_end and
      overlap == other.complement_overlap)
  end

  # Compares a link and optionally the complement link,
  # with two oriented_segments and optionally an overlap.
  # @param [RGFA::OrientedSegment] other_oriented_from
  # @param [RGFA::OrientedSegment] other_oriented_to
  # @param equivalent [Boolean] shall the complement link also be considered?
  # @param [RGFA::Alignment::CIGAR] other_overlap compared only if not empty
  # @return [Boolean] does the link or, if +equivalent+,
  #   the complement link go from the first
  #   oriented segment to the second with an overlap equal to the provided one
  #   (if not empty)?
  def compatible?(other_oriented_from, other_oriented_to, other_overlap = [],
                  equivalent = true)
    other_overlap = other_overlap.to_alignment
    is_direct = compatible_direct?(other_oriented_from, other_oriented_to,
                                   other_overlap)
    if is_direct
      return true
    elsif equivalent
      return compatible_complement?(other_oriented_from, other_oriented_to,
                          other_overlap)
    else
      return false
    end
  end

  # Compares a link with two oriented segments and optionally an overlap.
  # @param [RGFA::OrientedSegment] other_oriented_from
  # @param [RGFA::OrientedSegment] other_oriented_to
  # @param [RGFA::Alignment::CIGAR] other_overlap compared only if not empty
  # @return [Boolean] does the link go from the first
  #   oriented segment to the second with an overlap equal to the provided one
  #   (if not empty)?
  def compatible_direct?(other_oriented_from, other_oriented_to,
                         other_overlap = [])
    (oriented_from == other_oriented_from and
     oriented_to == other_oriented_to) and
     (overlap.empty? or other_overlap.empty? or (overlap == other_overlap))
  end

  # Compares the complement link with two oriented segments and optionally an
  # overlap.
  # @param [RGFA::OrientedSegment] other_oriented_from
  # @param [RGFA::OrientedSegment] other_oriented_to
  # @param [RGFA::Alignment::CIGAR] other_overlap compared only if not empty
  # @return [Boolean] does the complement link go from the first
  #   oriented segment to the second with an overlap equal to the provided one
  #   (if not empty)?
  def compatible_complement?(other_oriented_from, other_oriented_to,
                          other_overlap = [])
    (oriented_to == other_oriented_from.invert_orient and
     oriented_from == other_oriented_to.invert_orient) and
     (overlap.empty? or other_overlap.empty? or (overlap == other_overlap))
  end

  private

  def complement_ends?(other)
    (from_end == other.to_end and
      to_end == other.from_end)
  end

end

File: gfapy/line/edge/link/equivalence.py
import gfapy

class Equivalence:

  def __hash__(self):
    """
    Computes an hash for including the link in a dict,
    so that the hash of a link and its complement is the same.
    Thereby, tags are not considered.
    """
    hash(str(self.from_end)) + \
    hash(str(self.to_end)) + \
    hash(str(self.overlap)) + \
    hash(str(self.overlap.complement()))

  def is_eql(self, other):
    """
    Compares two links and determine their equivalence.
    Thereby, tags are not considered.

    .. note::
      Inverting the strand of both links and reversing
      the CIGAR operations (order/type), one obtains an
      equivalent complement link.

    Parameters
    ----------
    other : gfapy.line.edge.Link
      A link.

    Returns
    -------
    bool
      Are self and other equivalent?

    See Also
    --------
    ==
    is_same
    is_complement
    """
    return (self.is_same(other) or self.is_complement(other))

  def are_tags_eql(self, other):
    """
    Compares the tags of two links.

    .. note::
      This method shall be overridden if custom tags
      are defined, which have a complementation operation which determines
      their value in the equivalent but complement link.

    Parameters
    ----------
    other : gfapy.line.edge.Link
      A link.

    Returns
    -------
    bool
      Are self and other equivalent?

    See Also
    --------
    __eq__
    """
    return (sorted(self.tagnames) == sorted(other.tagnames)) and \
            all(self.get(fn) == other.get(fn) for fn in self.tagnames)

#  def __eq__(self, other):
#    """
#    Compares two links and determine their equivalence.
#    Tags must have the same content.
#
#    .. note:: Inverting the strand of both links and reversing
#    the CIGAR operations (order/type), one obtains an equivalent
#    link.
#
#    Parameters
#    ----------
#    other : gfapy.line.edge.Link
#      A link.
#
#    Returns
#    -------
#    bool
#      Are self and other equivalent?
#
#    See Also
#    --------
#    is_eql
#    are_tags_eql
#    """
#    return self.is_eql(other) and self.are_tags_eql(other)

  def is_same(self, other):
    """
    Compares two links and determine their equivalence.
    Thereby, tags are not considered.

    Parameters
    ----------
    other : gfapy.line.edge.Link
      A link.

    Returns
    -------
    bool
      Are self and other equivalent?

    See Also
    --------
    is_eql
    is_complement
    __eq__
    """
    return (self.from_end == other.from_end and
            self.to_end == other.to_end and
            self.overlap == other.overlap)

  def is_complement(self, other):
    """
    Compares the link to the complement of another link
    and determine their equivalence.
    Thereby, tags are not considered.

    Parameters
    ----------
    other : gfapy.line.edge.Link
      The other link.

    Returns
    -------
    bool
      Are self and the complement of other equivalent?

    See Also
    --------
    is_eql
    is_same
    __eq__
    """
    return (self.from_end == other.to_end and
            self.to_end == other.from_end and
            self.overlap == other.overlap.complement())

  def is_compatible(self, other_oriented_from, other_oriented_to,
                    other_overlap = None, allow_complement = True):
    """
    Compares a link and optionally the complement link,
    with two oriented_segments and optionally an overlap.

    Parameters
    ----------
    other_oriented_from : gfapy.OrientedLine
    other_oriented_to : gfapy.OrientedLine
    allow_complement : bool
      Shall the complement link also be considered?
    other_overlap : gfapy.Alignment.CIGAR
      Compared only if not empty.

    Returns
    -------
    bool
      Does the link or, if **allow_complement**, the complement link go from
      the first, oriented segment to the second with an overlap equal to the
      provided one (if not empty)?
    """
    other_overlap = gfapy.Alignment(other_overlap, version = "gfa1",
        valid = True)
    if self.is_compatible_direct(other_oriented_from, other_oriented_to,
        other_overlap):
      return True
    elif allow_complement:
      return self.is_compatible_complement(other_oriented_from,
                                           other_oriented_to,
                                           other_overlap)
    else:
      return False

  def is_compatible_direct(self, other_oriented_from, other_oriented_to,
                           other_overlap = None):
    """
    Compares a link with two oriented segments and optionally an overlap.

    Parameters
    ----------
    other_oriented_from : gfapy.OrientedLine
    other_oriented_to : gfapy.OrientedLine
    other_overlap : gfapy.Alignment.CIGAR
      Compared only if not empty.

    Returns
    -------
    bool
      Does the link go from the first oriented segment to the second
      with an overlap equal to the provided one (if not empty)?
    """
    return ((self.oriented_from == other_oriented_from and
             self.oriented_to == other_oriented_to) and
           (not self.overlap or not other_overlap or
             (self.overlap == other_overlap)))

  def is_compatible_complement(self, other_oriented_from, other_oriented_to,
                               other_overlap = None):
    """
    Compares the complement link with two oriented segments and optionally an
    overlap.

    Parameters
    ----------
    other_oriented_from : gfapy.OrientedLine
    other_oriented_to : gfapy.OrientedLine
    other_overlap : gfapy.Alignment.CIGAR
      Compared only if not empty.

    Returns
    -------
    bool
      Does the complement link go from the first oriented segment
      to the second with an overlap equal to the provided one (if not empty)?
    """
    return ((self.oriented_to == other_oriented_from.inverted() and
            (self.oriented_from == other_oriented_to.inverted()) and
            (not self.overlap or not other_overlap or
            (self.overlap == other_overlap.complement()))))

  def _complement_ends(self, other):
    return (self.from_end == other.to_end and self.to_end == other.from_end)
